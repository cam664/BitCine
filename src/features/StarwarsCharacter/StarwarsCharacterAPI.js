import fetch from 'isomorphic-fetch';

export const fetchPersonAPI = (id) => {
  return fetch(`https://swapi.co/api/people/${id}`);
}