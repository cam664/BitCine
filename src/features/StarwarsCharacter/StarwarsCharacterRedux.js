import { 
  fetchPersonAPI
} from './StarwarsCharacterAPI';

export const fetchPersonActions = {
  REQUEST: 'StarwarsCharacter/fetchPerson/REQUEST',
  SUCCESS: 'StarwarsCharacter/fetchPerson/SUCCESS',
  FAILURE: 'StarwarsCharacter/fetchPerson/FAILURE'
}

const initialState = {
  status: 'loading',
  data: {}
}

export default function reducer(state = initialState, action) {
  switch (action.type) {
    // fetchPerson
    case fetchPersonActions.REQUEST:
      return {
        ...state,
        status: 'loading',
        data: {},
      }
    case fetchPersonActions.SUCCESS:
      return {
        ...state,
        status: 'success',
        data: action.data
      }
    case fetchPersonActions.FAILURE:
      return {
        ...state,
        status: 'failure',
        data: {},
      }

    default:
      return state;
  }
}

export function fetchPerson(id) {
  return async (dispatch) => {
    dispatch({ type: fetchPersonActions.REQUEST });

    try {
      const res = await fetchPersonAPI(id).then(res => res.json());

      dispatch({
        type: fetchPersonActions.SUCCESS,
        data: res
      });
    } catch(e) {
      dispatch({ type: fetchPersonActions.FAILURE });
    }
  }
}