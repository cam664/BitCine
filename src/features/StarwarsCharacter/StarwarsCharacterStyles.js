export default {
  dl: `
    dt {
      font-weight: 700;
      margin-bottom: 4px;
      margin-top: 12px;
      &:first-of-type {
        margin-top: 0;
      }
    }
  `
}