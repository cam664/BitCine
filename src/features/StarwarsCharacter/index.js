export { default } from './StarwarsCharacter';

export { default as StarwarsCharacterReducer } from './StarwarsCharacterRedux';