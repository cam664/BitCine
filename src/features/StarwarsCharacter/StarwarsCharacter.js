import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { withRouter } from 'react-router-dom';

import Spinner from '../../common/Spinner';

import { fetchPerson } from './StarwarsCharacterRedux';

/** @jsx jsx */
import { css, jsx } from '@emotion/core';
import styles from './StarwarsCharacterStyles';

const StarwarsCharacter = (props) => {
  const dispatch = useDispatch();

  const characterData = useSelector(state => state.StarwarsCharacter);
  useEffect(() => {
    const characterID = props.location.pathname.split('/').pop();
    dispatch(fetchPerson(characterID));
  }, []);

  return (
    <>
      {characterData.status !== 'success' 
      ? (
        <Spinner />
      ) : (
        <article>
          <h1>{characterData.data.name}</h1>
          <dl css={css`${styles.dl}`}>
            <dt>Birth Year</dt>
            <dd>{characterData.data.birth_year}</dd>
            <dt>Eye Color</dt>
            <dd>{characterData.data.eye_color}</dd>
            <dt>Gender</dt>
            <dd>{characterData.data.gender}</dd>
            <dt>Hair Colour</dt>
            <dd>{characterData.data.hair_color}</dd>
            <dt>Mass</dt>
            <dd>{characterData.data.mass}</dd>
            <dt>Skin Colour</dt>
            <dd>{characterData.data.skin_color}</dd>
          </dl>
        </article>
      )}
    </>
  );
}

export default withRouter(StarwarsCharacter);