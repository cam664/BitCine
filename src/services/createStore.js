import { createStore, combineReducers, applyMiddleware } from "redux";
import thunkMiddleware from "redux-thunk";

import { StarwarsListReducer } from '../features/StarwarsList';

const reducer = combineReducers({
  StarwarsList: StarwarsListReducer,
});

export default (initialState) => 
  createStore(
    reducer, 
    initialState, 
    applyMiddleware(thunkMiddleware)
  );
