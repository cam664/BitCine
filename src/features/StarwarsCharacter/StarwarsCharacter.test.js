import React from 'react';
import renderer from 'react-test-renderer';
import { Provider as ReduxProvider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';

import StarwarsCharacter from './StarwarsCharacter';
import { fetchPersonActions } from './StarwarsCharacterRedux';
import { default as StarwarsCharacterReducer } from './StarwarsCharacterRedux';

import createStore from '../../services/createStore';

const store = createStore(window.__PRELOADED_STATE__);

describe('StarwarsCharacter', () => {
  it('renders a snapshot', () => {
    const component = renderer.create(
      <ReduxProvider store={store}>
        <Router>
          <StarwarsCharacter />
        </Router>
      </ReduxProvider>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('reducer', () => {
    const data = { name: 'John Smith' };

    it('should return a data object', () => {
      const state = { data: {}, status: 'loading' };
      const newState = StarwarsCharacterReducer(state, {
        type: fetchPersonActions.SUCCESS,
        data
      });
      expect(newState).toEqual({ data, status: 'success' });
    });

    it('should reset the error if data object is set', () => {
      const state = { data: {}, status: 'failure'};
      const newState = StarwarsCharacterReducer(state, {
        type: fetchPersonActions.SUCCESS,
        data
      });
      expect(newState).toEqual({ data, status: 'success' });
    });

    it('should set the error', () => {
      const state = { data, status: 'loading' };
      const newState = StarwarsCharacterReducer(state, {
        type: fetchPersonActions.FAILURE,
        data: {}
      });
      expect(newState).toEqual({ data: {}, status: 'failure' });
    });

  })
});