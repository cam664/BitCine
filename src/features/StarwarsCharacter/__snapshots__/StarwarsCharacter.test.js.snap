// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StarwarsCharacter renders a snapshot 1`] = `
<img
  className="css-14hden7-Spinner"
  src="spinner.gif"
/>
`;
